### PR TITLE
Pattern match improvements in order to move it down to the platform layer.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @grabalon

**Comment by CyrusNajmabadi** [comment]
@grabalon this contains several things we talked about at hte last meeting including:

1. Removing "GetMatchForLastSegmentOfPattern"
2. Making "IncludeMatchSpans" part of the matcher, not part of individual operations.
3. Having distinct 'kinds' for the different types of camel matches.

Note: i think Camel matches shoudl have even more kinds than what roslyn has defined.  I can enivsion the following all being useful:

1. Did the pattern match the first hump in the candidate?  i.e. matching "FB" against "FooBarQuux" should be better than "FB" against "QuuxFooBar".
2. Did the pattern match *all* humps in the pattern.  i.e. matching "FB" against "FooBar" should be better than "FB" against "FooBarQuux".
3. Did the pattern match humps contiguously?  Or was there a gap anywhere?  i.e. matching "FB" against "QuuxFooBarZtesch" should likely be better than matching "FB" agaisnt "QuuxFooZteschBar".  

Right now roslyn has these concepts *except* for '2'.  Matching all the humps should be taken into account when weighting the patterns.

**Comment by CyrusNajmabadi** [comment]
tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
@grabalon I'm going to work on making it so that the API does not return arrays and instead is always given allocating contexts to place the result into.

**Comment by CyrusNajmabadi** [comment]
@grabalon I've also moved the API over to using "accumulators" to store results in instead of returning full arrays.  Note: i'm using ArrayBuilder because that's the normal roslyn type we use here.  We'd likely want some new cleaner type with less surface area in the core API.

Also, i'm thinking we want to simplify this API even more.  i.e. we could create a pattern matcher for something like "S.C.WL", but then tehre would be two entrypoints.  One that says "does this simple name match the simple name part of the pattern?" and one that says "does this container match the container part of the pattern?" (and potentially an API that combines both).

**Comment by CyrusNajmabadi** [comment]
@grabalon I'm thinking it might be a good idea to make the "container matching" system more flexible/uniform.  I'm thinking about something like so:  (note: names are terrible):

```c#

PatternMatcher(string[] patternParts, char[] candidateSplitChars, ...)

Match(string candidateToBeSplit)
Match(string candidateToBeSplit, string candidateToBeSplit2)
Match(string candidateToBeSplit1, string candidateToBeSplit2, string candidateToBeSplit3)
Match(string[] candidatesToBeSplit)
```

The idea here is that we would use the "candidateSplitChars" to actually "virtually" break up these candidates, and we would then match the pattern against them from right to left.  This would allow clients to flexibly store things in their indices, while still getting this sort of hierarchical pattern matching.

For example, while ML stores our symbols as "Fully.Qualified.Container"+"SimpleName", other languages lmight just store the full name like "Fully::Qualified.Container.SimpleName" or "Fully"+"Qualified"+"Container"+"SimpleName".  

The reason for the overloads is to prevent actual string allocations for the common cases of index values being passed in.  The API would *act* as if it was actually splitting the strings with the given "candidateSplitStrings", but under the covers it would be efficiently doing the matching without actually doing that.

--

Interestingly enough, ML never uses the pattern match spans for container matches.  So we could even have two APIs, one that is just for testing a simple name against a simple pattern, and one which does this sort of virtual candidate breaking and matches against a list of patterns.

This helps simplify things greatly because we don't need to figure out how to give back the matching spans and relate it to the arguments for a method like ```Match(string candidateToBeSplit1, string candidateToBeSplit2, string candidateToBeSplit3)```.

--

**Comment by CyrusNajmabadi** [comment]
Tagging @ManishJayaswal 

These are a set of improvements to an API Roslyn has that Platform would like to move down so that it's available to anyone building on top of our core VS APIs.

File: src/EditorFeatures/Test/Utilities/PatternMatcherTests.cs [hunk 1]
```diff
@@ -199,45 +199,45 @@ private void VerifyBreakIntoCharacterParts(string original, params string[] part
...
+        [InlineData("[|co|]deFix[|Pro|]vider", "copro", PatternMatchKind.CamelCaseFromStart, CaseInsensitive)]
+        [InlineData("[|Co|]deFix_[|Pro|]vider", "copro", PatternMatchKind.CamelCaseFromStart, CaseInsensitive)]
+        [InlineData("[|C|]ore[|Ofi|]lac[|Pro|]fessional", "cofipro", PatternMatchKind.CamelCaseContiguousFromStart, CaseInsensitive)]
+        [InlineData("[|C|]lear[|Ofi|]lac[|Pro|]fessional", "cofipro", PatternMatchKind.CamelCaseContiguousFromStart, CaseInsensitive)]
+        [InlineData("[|CO|]DE_FIX_[|PRO|]VIDER", "copro", PatternMatchKind.CamelCaseFromStart, CaseInsensitive)]
+
+        [InlineData("my[|_b|]utton", "_B", PatternMatchKind.CamelCaseContiguous, CaseInsensitive)]
+        [InlineData("[|_|]my_[|b|]utton", "_B", PatternMatchKind.CamelCaseFromStart, CaseInsensitive)]
         public void TestGetFirstMatch(
             string candidate, string pattern, int matchKindInt, bool isCaseSensitive, int? camelCaseWeight = null)
```
Comment (sharwell) [inline src/EditorFeatures/Test/Utilities/PatternMatcherTests.cs]: :bulb: It appears `camelCaseWeight` is no longer used by any of the test cases.

**Comment by CyrusNajmabadi** [comment]
Tagging @JoshVarty

**Comment by CyrusNajmabadi** [comment]
Closing in favor of https://github.com/dotnet/roslyn/pull/18741

**Pull request closed without merging**